cli/src/commands/operation/diff.rs

```diff
@@ -0,0 +1,527 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use indexmap::IndexMap;
+use itertools::Itertools;
+use jj_lib::backend::{BackendResult, ChangeId, CommitId};
+use jj_lib::commit::Commit;
+use jj_lib::git::REMOTE_NAME_FOR_LOCAL_GIT_REPO;
+use jj_lib::graph::{GraphEdge, TopoGroupedGraphIterator};
+use jj_lib::matchers::EverythingMatcher;
+use jj_lib::op_store::{RefTarget, RemoteRef, RemoteRefState};
+use jj_lib::refs::{diff_named_ref_targets, diff_named_remote_refs};
+use jj_lib::repo::{MutableRepo, ReadonlyRepo, Repo};
+use jj_lib::revset::RevsetIteratorExt as _;
+use jj_lib::rewrite::rebase_to_dest_parent;
+use jj_lib::{dag_walk, op_walk, revset};
+
+use crate::cli_util::{
+    short_change_hash, short_operation_hash, CommandHelper, LogContentFormat,
+    WorkspaceCommandTransaction,
+};
+use crate::command_error::{user_error, CommandError};
+use crate::diff_util::{DiffFormatArgs, DiffRenderer};
+use crate::formatter::Formatter;
+use crate::graphlog::{get_graphlog, Edge};
+use crate::ui::Ui;
+
+/// Compare changes to the repository between two operations
+#[derive(clap::Args, Clone, Debug)]
+pub struct OperationDiffArgs {
+    /// Show repository changes in this operation, compared to its parent
+    #[arg(long, visible_alias = "op")]
+    operation: Option<String>,
+    /// Show repository changes from this operation
+    #[arg(long, conflicts_with = "operation")]
+    from: Option<String>,
+    /// Show repository changes to this operation
+    #[arg(long, conflicts_with = "operation")]
+    to: Option<String>,
+    /// Don't show the graph, show a flat list of modified changes
+    #[arg(long)]
+    no_graph: bool,
+    /// Show patch of modifications to changes
+    ///
+    /// If the previous version has different parents, it will be temporarily
+    /// rebased to the parents of the new version, so the diff is not
+    /// contaminated by unrelated changes.
+    #[arg(long, short = 'p')]
+    patch: bool,
+    #[command(flatten)]
+    diff_format: DiffFormatArgs,
+}
+
+pub fn cmd_op_diff(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &OperationDiffArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
```

Comment:
nit: `workspace = command.load_workspace()` can be moved here. RepoLoader can be obtained from the workspace.

Reply:
Hmm, if I use `command.load_workspace()`, does that mean we don't want to perform snapshotting (which is done by `command.workspace_helper(ui)`)?

Reply:
I don't mind if `op diff` doesn't trigger snapshot. There's no reason to avoid snapshotting here, but it's not needed either.

Reply:
I've kept the current snapshotting for now, so I don't think this comment is still applicable.